packages/babel-plugin-transform-typescript/src/index.js

```diff
@@ -0,0 +1,255 @@
+import syntaxTypeScript from "babel-plugin-syntax-typescript";
+
+import transpileEnum from "./enum";
+
+function isInType(path) {
+  switch (path.parent.type) {
+    case "TSTypeReference":
+    case "TSQualifiedName":
+    case "TSExpressionWithTypeArguments":
+    case "TSTypeQuery":
+      return true;
+    default:
+      return false;
+  }
+}
+
+interface State {
+  programPath: any,
+}
+
+export default function({ types: t }) {
+  return {
+    inherits: syntaxTypeScript,
+    visitor: {
+      //"Pattern" alias doesn't include Identifier or RestElement.
+      Pattern: visitPattern,
+      Identifier: visitPattern,
+      RestElement: visitPattern,
+
+      Program(path, state: State) {
+        state.programPath = path;
+      },
+
+      ImportDeclaration(path, state: State) {
+        // Note: this will allow both `import { } from "m"` and `import "m";`.
+        // In TypeScript, the former would be elided.
+        if (path.node.specifiers.length === 0) {
+          return;
+        }
+
+        let allElided = true;
+        const importsToRemove: Path<Node>[] = [];
+
+        for (const specifier of path.node.specifiers) {
+          const binding = path.scope.getBinding(specifier.local.name);
+          if (isImportTypeOnly(binding, state.programPath)) {
+            importsToRemove.push(binding.path);
+          } else {
+            allElided = false;
+          }
+        }
+
+        if (allElided) {
+          path.remove();
+        } else {
+          for (const importPath of importsToRemove) {
+            importPath.remove();
+          }
+        }
+      },
+
+      TSDeclareFunction(path) {
```

Comment:
The only ones which can be directly removed are `TSDeclareFunction`, `TSDeclareMethod`, `TSIndexSignature`, `TSInterfaceDeclaration`, and `TSTypeAliasDeclaration`.
What these all have in common is that they are statements, and are always *just* declarations (no emit like an `enum` has).

Nodes like `TSAsExpression` can't be entirely removed, and nodes like `TSModuleDeclaration` requires special handling depending on whether it's a declaration.

If we just added a `TypeScript` alias it would be confusing why that doesn't include anything else. I could make a `TSNonValueDeclaration` alias though?